packages/salesforcedx-vscode-replay-debugger/src/breakpoints/checkpointService.ts

```diff
@@ -193,5 +426,112 @@ export class CheckpointInfoIterationNode extends CheckpointInfoNode {
+// The file location that matches our URI from the SourceBreakpoint is the encoded URI
+// string from Location.Uri.toString(false /*skipEncoding*/). It'll have same format as the
+// URI is the setBreakPointsRequest in apexReplayDebug.ts. The line number is also in the
+// location and started as a range. Pull the line number from Location.Range.Start.Line but
+// add 1 because those lines are 0 based.
+// breakpointsChangedEvent.added:
+// This event fires with items in the added array when the extensions are first launched
+// if there were breakpoints in the loaded project. The checkpoints will be entirely empty
+// at that point and time so they need to be populated.
+// breakpointsChangedEvent.changed:
+// The event fires with a list of changed breakpoint events at project shutdown even though
+// nothing in those events has actually changed. Currently the changed event is ignored.
+// breakpointsChangedEvent.removed:
+// The removed events are only going to fire when something is actually removed so unlike
+// the added event there's no ambiguity there
+export async function processBreakpointChangedForCheckpoints(
+  breakpointsChangedEvent: vscode.BreakpointsChangeEvent
+): Promise<void> {
+  const breakpointsToRemove: vscode.Breakpoint[] = [];
+  for (const bp of breakpointsChangedEvent.added) {
+    if (
+      bp.enabled &&
+      bp.condition &&
+      bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0
+    ) {
+      if (!checkpointService.canAddCheckpointNote()) {
+        breakpointsToRemove.push(bp);
+        continue; // go on to the next iteration
+      }
+      if (bp instanceof vscode.SourceBreakpoint) {
+        const uri = bp.location.uri.toString(false /* skipEncoding */);
+        const typeRef = BreakpointUtil.getInstance().getTopLevelTyperefForUri(
+          uri
+        );
+        const filename = uri.substring(uri.lastIndexOf('/') + 1);
+        await checkpointService.addCheckpointNode(
+          uri,
+          filename,
+          typeRef,
+          bp.location.range.start.line + 1 // need to add 1 since the lines are 0 based
+        );
+      }
+    }
+  }
+  // breakpointsToRemove will be empty unless the user tried to add more than 5 checkpoints.
+  // Call showErrorMessage to dispaly the error message once and remove the breakpoints.
+  // Unfortunately, calling delete below is going to mean that this event handler is going to
+  // get called for the delete.
+  if (breakpointsToRemove.length > 0) {
+    vscode.window.showErrorMessage(nls.localize('up_to_five_checkpoints'));
+    vscode.debug.removeBreakpoints(breakpointsToRemove);
+  }
+
+  // The remove always needs to call delete to get rid of the checkpoint.
+  for (const bp of breakpointsChangedEvent.removed) {
+    if (
+      bp.enabled &&
+      bp.condition &&
+      bp.condition!.toLowerCase().indexOf(CHECKPOINT) >= 0
+    ) {
+      if (bp instanceof vscode.SourceBreakpoint) {
```

Comment:
I think deleting the checkpoints should happen before adding them since there is a limit of 5 checkpoints.

Reply:
Yeah, I'll move this but funny story: The only time any of these arrays contains more than one object is when VS Code first loads the project and is loading beakpoints which is specific to add. Otherwise, from the UI they happen one at a time.